### ISSUE TITLE
EC pretraining model generates short sentences and has poor accuracy

### ISSUE THREAD

**Issue by XuhuiZhou** [cambridgeltl/ECNMT#1]
Hi, I am re-running the ECPRETRAIN script and trying to reproduce the results.

However, I found the model tends to generate very short sentences as the training goes on:
![image](https://user-images.githubusercontent.com/20436061/102169156-6b136c80-3e46-11eb-9f91-04317daccace.png)
This seems to be not aligned with the description in the paper.

Besides, it seems that we only save a model when it reaches 99% accuracy, however, none of the training epoch reaches that in my experiment.

I try to reproduce the environment, however, I found that ```Pytorch 1.3.1``` does not seem to be available on Pytorch's official webpage anymore, so I use ```Pytorch 1.6``` instead. I do not know whether this could be the reason. Any insight here would be helpful.

**Comment by yaoyiran** [comment]
Hi, can you lmk your hyper-parameters and your data? In our experiments, the prediction accuracy can easily achieve >99.5% at only epoch 4000 or so. I do not know if there would be any issue caused by different pytorch versions but please make sure that you could follow our original settings especially for those hyperparameters. I would recommend you to run the codes again and see if you could get different results. We use nvidia-docker so we are able to download official images with any pytorch versions which we use to build containers.
<img width="564" alt="Screenshot 2020-12-17 at 10 57 43" src="https://user-images.githubusercontent.com/25559589/102479390-bf8d3680-4056-11eb-86ba-1b57b1ef967c.png">

**Comment by yaoyiran** [comment]
I downloaded our Github code myself and re-run it again with all its default settings (In pytorch 1.3) this afternoon and I confirm that the valid prediction accuracy can easily achieve >99% at epoch 3000~4000. In addition, the average message length usually keeps being 15. If you need the log from this run, I can share it with you. 

If you would like to install pytorch 1.3.1, here might be a possible way:
```bash
pip install --upgrade pip
pip install torch==1.3.1
```

Currently, we do not understand the influence of different pytorch versions. We would appreciate it if you could identify any and let us know.

**Comment by XuhuiZhou** [comment]
Hi, thanks for the suggestion!
I also obtained the 99% prediction accuracy by re-running the experiments, so it is not a problem with PyTorch's version but more like some random states cause the divergence of the model.

BTW, I am using the default ```./run_training.sh``` and I double-checked the default hyperparameters, which seems to be the same as the paper. I also downloaded the data from the repo. So the setting should be the same as your experiments.

One question, I notice that the default training epoch is around 30k. Since the valid results tend to be perfect around 3k epoch, do you intentionally train them longer? And how you choose the model? The best on the valid set? (I might miss this detail in the paper)

**Comment by yaoyiran** [comment]
Obviously, the best way to select EC pretrained model is to evaluate each saved pretrained model on your downstream task (using its dev set only). You could even run the pretraining procedure for several times and saved as many models as possible. However, if you do not have time, you can just pick up the first model that achieve some reasonable EC prediction accuracy (say 99%) which will still give you some good result on the downstream task.

Thanks for your interest in our project!
